Added test case to run VM Backup with freeze only without unfreeze and vice versa

mkoppal-px opened the pull request:
**What this PR does / why we need it**:

- [x] Added new test case `KubevirtVMWithFreezeUnfreeze`
- [x] Changed the testrail ID for `KubevirtVMBackupRestoreWithDifferentStates` and `KubevirtUpgradeTest`
- [x] Added new function to create freeze and unfreeze rules - `CreateVMRule`
- [x] Handled the panic in `RunCmdInVM` when the username or password is not found


**Which issue(s) this PR fixes** (optional)
Closes #PB-5790

**Special notes for your reviewer**:
[Jenkins Run on Vanilla](https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/4744/)

atrivedi-px commented on tests/backup_helper.go:
```diff
@@ -4920,6 +4948,41 @@ func CreateRuleForBackupWithMultipleApplications(orgID string, appList []string,
+	Unfreeze VMRuleType = "unfreeze"
+)
+
+func CreateVMRule(ruleName string, vms []kubevirtv1.VirtualMachine, ruleType VMRuleType, ctx context1.Context) error {
```
Could we use CreateRuleForKubevirtBackup from backup for this?

mkoppal-px replied:
I think we can call it `CreateRuleForVMBackup` instead of `Kubevirt` just be consistent. Is it OK?

mkoppal-px replied:
Done

kshithijiyer-px commented on tests/backup_helper.go:
```diff
@@ -4920,6 +4948,41 @@ func CreateRuleForBackupWithMultipleApplications(orgID string, appList []string,
+	Unfreeze VMRuleType = "unfreeze"
+)
+
+func CreateRuleForVMBackup(ruleName string, vms []kubevirtv1.VirtualMachine, ruleType VMRuleType, ctx context1.Context) error {
```
Can you add a doc sting here?

mkoppal-px replied:
Done

kshithijiyer-px commented on tests/backup_helper.go:
```diff
@@ -4920,6 +4948,41 @@ func CreateRuleForBackupWithMultipleApplications(orgID string, appList []string,
+	Unfreeze VMRuleType = "unfreeze"
+)
+
+func CreateRuleForVMBackup(ruleName string, vms []kubevirtv1.VirtualMachine, ruleType VMRuleType, ctx context1.Context) error {
```
Why can't we have the rukes as per the name of the vm?

mkoppal-px replied:
You can give any name to the rule. There are cases where there are VMs with same name in different namespaces.
So it is left to the caller to give any name required for the test case.

kshithijiyer-px commented on tests/backup_helper.go:
```diff
@@ -4920,6 +4948,41 @@ func CreateRuleForBackupWithMultipleApplications(orgID string, appList []string,
+			})
+		}
+	}
+	RuleCreateReq := &api.RuleCreateRequest{
```
Can't we use the existing createRule function?

mkoppal-px replied:
The create rule function that we have today is app specific, but he VM rule is not specific to any VM. Every VM will have the same rule. Only the VM name and namespace will change.

kshithijiyer-px replied:
@mkoppal-px can we move this to a function and call it at two places?

mkoppal-px replied:
Did not get you

mkoppal-px replied:
We can take this up later

kshithijiyer-px commented on tests/backup_helper.go:
```diff
@@ -4920,6 +4948,41 @@ func CreateRuleForBackupWithMultipleApplications(orgID string, appList []string,
 
+type VMRuleType string
+
+const (
```
You'll need anyways both for any given Vm right why do you want to specific VMRuleType in general?

mkoppal-px replied:
No, we need not have both for a VM. In my case I am creating only freeze rule and checking how backup works when I give only freeze rule without unfreeze.
If anybody needs a function with both freeze and unfreeze, then they can write a wrapper function which calls
```
err = CreateRuleForVMBackup(freezeRuleName, allVMs, Freeze, ctx)
err = CreateRuleForVMBackup(freezeRuleName, allVMs, Unfreeze, ctx)
```

kshithijiyer-px replied:
Got it.

kshithijiyer-px commented on tests/backup/backup_kubevirt_test.go:
```diff
@@ -1335,3 +1335,217 @@ var _ = Describe("{KubevirtVMBackupRestoreWithNodeSelector}", func() {
+			dash.VerifyFatal(clusterStatus, api.ClusterInfo_StatusInfo_Online, fmt.Sprintf("Verifying if [%s] cluster is online", SourceClusterName))
+
+			sourceClusterUid, err = Inst().Backup.GetClusterUID(ctx, BackupOrgID, SourceClusterName)
+			log.Infof(sourceClusterUid)
```
DO you really need this log line?

mkoppal-px replied:
Will remove it

kshithijiyer-px commented on tests/backup/backup_kubevirt_test.go:
```diff
@@ -1335,3 +1335,217 @@ var _ = Describe("{KubevirtVMBackupRestoreWithNodeSelector}", func() {
+		}
+
+		log.Info("Destroying scheduled apps on source cluster")
+		//err = DestroyAppsWithData(scheduledAppContexts, opts, controlChannel, errorGroup)
```
Do you really need this code?

mkoppal-px replied:
Yes, currently we do not have the VM data validation part ready. Once we have that I have to replace `ValidateApplications` with `DestroyAppsWithData`

mkoppal-px replied:
Yes

kshithijiyer-px commented on tests/backup/backup_kubevirt_test.go:
```diff
@@ -1335,3 +1335,217 @@ var _ = Describe("{KubevirtVMBackupRestoreWithNodeSelector}", func() {
+			err = CreateApplicationClusters(BackupOrgID, "", "", ctx)
+			dash.VerifyFatal(err, nil, "Creating source and destination cluster")
+
+			clusterStatus, err := Inst().Backup.GetClusterStatus(BackupOrgID, SourceClusterName, ctx)
```
Can we add a loop here where we run the same function calls for SourceClusterName and DestinationClusterName cluster name? 

```
for clusterName in [DestinationClusterName, SourceClusterName ]{
	clusterStatus, err = Inst().Backup.GetClusterStatus(BackupOrgID, clusterName, ctx)
			log.FailOnError(err, fmt.Sprintf("Fetching [%s] cluster status", clusterName))
			dash.VerifyFatal(clusterStatus, api.ClusterInfo_StatusInfo_Online, fmt.Sprintf("Verifying if [%s] cluster is online", clusterName))

}
```

mkoppal-px replied:
This is a good suggestion. Will add it. But it is there in almost all the tests.

kshithijiyer-px replied:
We need to clean such things at some point.

mkoppal-px replied:
Done

kshithijiyer-px reviewed (changes requested):
I see a few issues here can you take a look?

kphalgun-px commented on tests/backup/backup_kubevirt_test.go:
```diff
@@ -1335,3 +1335,217 @@ var _ = Describe("{KubevirtVMBackupRestoreWithNodeSelector}", func() {
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+			for _, vm := range allVMs {
+				log.Infof("Running command for VM [%s]", vm.Name)
+				output, err := RunCmdInVM(vm, "uname -a", ctx)
```
How about? 

```go
import "context"

output, err := RunCmdInVM(vm, "uname -a", context.TODO())
```

It will make it easy to reuse the code in other tests.

mkoppal-px replied:
Done

kphalgun-px commented on tests/backup_helper.go:
```diff
@@ -884,6 +899,19 @@ func GetNamespacesFromVMs(vms []kubevirtv1.VirtualMachine) []string {
+func GetNamespacesToVMsMap(vms []kubevirtv1.VirtualMachine) map[string][]string {
+	namespacesToVMsMap := make(map[string][]string)
+	for _, v := range vms {
+		if _, ok := namespacesToVMsMap[v.Namespace]; !ok {
+			namespacesToVMsMap[v.Namespace] = make([]string, 0)
+		}
```
I think there is no need to initialize slices; append will do the initialization if required.

mkoppal-px replied:
Done

kphalgun-px commented on tests/backup_helper.go:
```diff
@@ -6311,7 +6386,7 @@ func RunCmdInVM(vm kubevirtv1.VirtualMachine, cmd string, ctx context1.Context)
 		// To check if the ssh server is up and running
 		t := func() (interface{}, bool, error) {
-			output, err := k8sCore.RunCommandInPod(testCmdArgs, sshPodName, "ssh-container", "default")
+			output, err := k8sCore.RunCommandInPod(testCmdArgs, sshPodName, "ssh-container", sshPodNamespace)
```
👍

kphalgun-px commented on tests/backup_helper.go:
```diff
@@ -6282,17 +6347,27 @@ func RunCmdInVM(vm kubevirtv1.VirtualMachine, cmd string, ctx context1.Context)
 	}
-	password := cm.Data[vm.Name]
+	if p, ok := cm.Data[vm.Name]; !ok {
+		return "", fmt.Errorf("password not found in the configmap for vm - [%s]", vm.Name)
```
How about?

```go
return "", fmt.Errorf("password not found in the configmap [%s/%s] for vm - [%s]", "default", "kubevirt-creds", vm.Name)
```

kshithijiyer-px commented on tests/backup/backup_kubevirt_test.go:
```diff
@@ -1335,3 +1335,217 @@ var _ = Describe("{KubevirtVMBackupRestoreWithNodeSelector}", func() {
+		backupLocationMap    map[string]string
+		providers            []string
+		labelSelectors       map[string]string
+		//controlChannel       chan string
```
Can you move all the commented code in a single block and add a TODO on top? So that we don't lose track?

mkoppal-px replied:
Makes sense

kshithijiyer-px commented on tests/backup/backup_kubevirt_test.go:
```diff
@@ -1335,3 +1335,217 @@ var _ = Describe("{KubevirtVMBackupRestoreWithNodeSelector}", func() {
+
+		Step("Validating applications", func() {
+			log.InfoD("Validating applications")
+			//ctx, err := backup.GetAdminCtxFromSecret()
```
Do you really need it if not them remove it else add a comment with a TODO?

kphalgun-px reviewed (approved):
LGTM

mkoppal-px commented:
Latest run after addressing all review comments with OCP cluster - https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/4773/

aravindranath14 reviewed (approved):
LGTM

dbhatnagar-px reviewed (approved):
Lgtm
